File: src/py_pkg/py_pkg/action_client.py
import rclpy
from rclpy.action import ActionClient
from rclpy.node import Node
from example_interfaces.action import Fibonacci


class FibonacciActionClient(Node):
    def __init__(self):
        super().__init__("py_action_client")
        self._client = ActionClient(self, Fibonacci, "benchmark_fibonacci")

    def send_goal(self, order):
        goal = Fibonacci.Goal()
        goal.order = order
        self._client.wait_for_server()

        future = self._client.send_goal_async(
            goal, feedback_callback=self.feedback_callback
        )
        rclpy.spin_until_future_complete(self, future)
        goal_handle = future.result()

        if not goal_handle.accepted:
            self.get_logger().info("Goal rejected")
            return

        result_future = goal_handle.get_result_async()
        rclpy.spin_until_future_complete(self, result_future)
        result = result_future.result().result
        self.get_logger().info(f"Result: {len(result.sequence)} elements")

    def feedback_callback(self, feedback_msg):
        seq = feedback_msg.feedback.sequence
        if seq:
            self.get_logger().info(f"Feedback: len={len(seq)} last={seq[-1]}")


def main(args=None):
    rclpy.init(args=args)
    node = FibonacciActionClient()
    node.send_goal(10)
    node.destroy_node()
    rclpy.shutdown()

File: src/cpp_pkg/src/action_client.cpp
#include <chrono>
#include <iostream>
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "example_interfaces/action/fibonacci.hpp"

using Fibonacci = example_interfaces::action::Fibonacci;

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    auto node = rclcpp::Node::make_shared("cpp_action_client");
    auto client = rclcpp_action::create_client<Fibonacci>(node, "benchmark_fibonacci");

    if (!client->wait_for_action_server(std::chrono::seconds(5))) {
        RCLCPP_ERROR(node->get_logger(), "Action server not available");
        return 1;
    }

    auto goal = Fibonacci::Goal();
    goal.order = 10;

    auto send_goal_options = rclcpp_action::Client<Fibonacci>::SendGoalOptions();
    send_goal_options.feedback_callback =
        [&node](auto, const auto& feedback) {
            auto& seq = feedback->sequence;
            if (!seq.empty()) {
                RCLCPP_INFO(node->get_logger(), "Feedback: len=%zu last=%d",
                    seq.size(), seq.back());
            }
        };

    auto goal_future = client->async_send_goal(goal, send_goal_options);
    if (rclcpp::spin_until_future_complete(node, goal_future) != rclcpp::FutureReturnCode::SUCCESS) {
        RCLCPP_ERROR(node->get_logger(), "Failed to send goal");
        return 1;
    }

    auto goal_handle = goal_future.get();
    if (!goal_handle) {
        RCLCPP_ERROR(node->get_logger(), "Goal was rejected");
        return 1;
    }

    auto result_future = client->async_get_result(goal_handle);
    if (rclcpp::spin_until_future_complete(node, result_future) == rclcpp::FutureReturnCode::SUCCESS) {
        auto& seq = result_future.get().result->sequence;
        RCLCPP_INFO(node->get_logger(), "Result: %zu elements", seq.size());
    }

    rclcpp::shutdown();
    return 0;
}
